src/lava/lib/optimization/solvers/generic/monitoring_processes/solution_readout/models.py

```diff
@@ -45,7 +47,10 @@ def run_spk(self):
             timestep, raw_solution = self._receive_data()
             cost = self.decode_cost(raw_cost)
             self.solution_step = abs(timestep)
-            self.solution[:] = self.decode_solution(raw_solution)
+            self.solution[:] = self.decode_solution(
+                raw_solution=raw_solution,
+                time_steps_per_algorithmic_step=self.
+                time_steps_per_algorithmic_step)

```

Comment:
This is where things can potentially go wrong with class variables. `self.time_steps_per_algorithmic_step` is not bound to this instance. If someone changes `<classname>.time_steps_per_algorithmic_step` then its value will change for all instances.

```python
class A:
    class_var = 5
    def __init__(self, val):
        inst_var = val

>> a_instance = A(13)  # this creates an instance with a_instance.class_var = 5 and a_instance.inst_var = 13
>> A.class_var = 7  # this changes class_var even for a_instance

Reply:
Hm. Could you elaborate?

Please note that in general, I used <Process>.time_steps_per_algorithmic_step as property of the Process for DiscreteVariable-Neurons, like SimulatedAnnealing or StochasticIntegrateAndFire. There, they are supposed to be fixed properties, that must not be altered by the user. I could prevent this from happening altogether by setting the __setattr__ method of the class; but my guess was that no user will ever figure out how to access these Processes in the convoluted OptimizationSolver, anyway.

In the line that you comment on here, the OptimizationSolver has provided the value of <Process>.time_steps_per_algorithmic_step to the SolutionReadout, which has defined it as Var. Here, it is not a class variable anymore, but a variable of this instance of the SolutionReadout Process!

It still should not be changed after initiation, that's true. Thus, I could use getter/setters to prevent this. But at the moment, I'm uncertain whether the compiler understands what happens if we set the Var of a Process using a Getter/Setter.

But maybe you could elaborate your thoughts?

Reply:
My issue was with making it a class variable than a private variable of an instance, guarded by a getter/setter. But Python also has the philosophy that developers need to be trusted to do the right thing. So we can leave this as it is.